libtransmission/tr-dht.cc

```diff
@@ -41,498 +38,376 @@
+        if (swarmStatus(AF_INET) < SwarmStatus::Poor && swarmStatus(AF_INET6) < SwarmStatus::Poor)
+        {
+            return;
+        }
+
+        auto const now = tr_time();
+        for (auto const id : mediator_.torrentsAllowingDHT())
         {
-            /* We don't want to abuse our bootstrap nodes, so be very
-               slow.  The initial wait is to give other nodes a chance
-               to contact us before we attempt to contact a bootstrap
-               node, for example because we've just been restarted. */
-            nap(40);
+            auto& times = announce_times_[id];
 
-            if (isBootstrapDone())
+            if (auto& announce_after = times.ipv4_announce_after; announce_after < now)
             {
-                break;
+                auto const announce_again_in_n_secs = announceTorrent(mediator_.torrentInfoHash(id), AF_INET, peer_port_);
+                announce_after = now + std::chrono::seconds{ announce_again_in_n_secs }.count();
             }
 
-            if (i == 0)
+            if (auto& announce_after = times.ipv6_announce_after; announce_after < now)
             {
-                tr_logAddDebug("Attempting bootstrap from dht.transmissionbt.com");
+                auto const announce_again_in_n_secs = announceTorrent(mediator_.torrentInfoHash(id), AF_INET6, peer_port_);
+                announce_after = now + std::chrono::seconds{ announce_again_in_n_secs }.count();
             }
-
-            bootstrapFromName("dht.transmissionbt.com", tr_port::fromHost(6881), getBootstrappedAF());
         }
     }
 
-    tr_logAddTrace("Finished bootstrapping");
-}
+    ///
 
-int tr_dhtInit(tr_session* session, tr_socket_t udp4_socket, tr_socket_t udp6_socket)
-{
-    if (impl.session != nullptr) /* already initialized */
+    void onPeriodicTimer()
     {
-        return -1;
-    }
+        auto const call_again_in_n_secs = periodic(nullptr, 0, nullptr, 0);
 
-    tr_logAddInfo(_("Initializing DHT"));
+        // Being slightly late is fine,
+        // and has the added benefit of adding some jitter.
+        auto const interval = call_again_in_n_secs + std::chrono::milliseconds{ tr_rand_int_weak(1000) };
+        periodic_timer_->startSingleShot(interval);
+    }
 
-    if (tr_env_key_exists("TR_DHT_VERBOSE"))
+    [[nodiscard]] std::chrono::seconds periodic(void const* buf, size_t buflen, struct sockaddr const* from, socklen_t fromlen)
     {
-        dht_debug = stderr;
-    }
+        auto call_again_in_n_secs = time_t{};
 
-    auto benc = tr_variant{};
-    auto const dat_file = tr_pathbuf{ session->configDir(), "/dht.dat"sv };
-    auto const ok = tr_variantFromFile(&benc, TR_VARIANT_PARSE_BENC, dat_file.sv());
+        // the only way dht_periodic() fails is if buf isn't zero-terminated,
+        // so let's ensure it's zero terminated here.
+        auto szbuf = tr_strbuf<unsigned char, 1500>{ std::string_view{ static_cast<char const*>(buf), buflen } };
```

Comment:
This was previously handled by an `buf[buflen] = '\0'`:

https://github.com/transmission/transmission/blob/69fda72a8b44bde0cc6b1842f79d5a0a23cc4321/libtransmission/tr-udp.cc#L212

The arbitrary value of 1500 seems like a possible source of bugs or performance issues.

Reply:
:+1: 

OT1H, it's not going to cause any bugs. `tr_strbuf` automatically resizes itself; the `1500` is how much gets allocated on the stack to avoid expensive malloc / free calls.

OT2H, the old approach was slightly more efficient since we didn't need to make a temporary copy just to ensure zero termination. I've gone back to the old approach in e88716dd96699acf94dd67193c456edf06f866e2